Experimental new client API.

petermattis opened the pull request:
Not finished, but ready for comments.

petermattis committed:
Experimental new client API.

spencerkimball commented on client/db.go:
```diff
@@ -0,0 +1,461 @@
+
+// Timestamp ...
+type Timestamp struct {
+	WallTime int64
```
I'd strip out logical time from end client timestamp. It's utterly worthless from client perspective, or worse, confusing.

petermattis replied:
Ok, I was wondering about that. Removal simplifies this down to an `int64` which is easily converted to a time.Time.

bdarnell commented on client/db_test.go:
```diff
@@ -0,0 +1,28 @@
+	result, _ := db.Get("hello")
+	fmt.Println(result.Rows[0].ValueString())
+
+	db.Inc(100, "world")
```
`Inc(value, key)` feels really unnatural, especially in comparison to `Put(key, value)`. I understand that this is the only way we can get variadic increments, but do we really need that?

For that matter I'm not a fan of variadic Put() either, since we have explicit batches. I think it would be cleaner to just call Batch.Put multiple times even if it is more verbose.

petermattis replied:
Yeah, the variadic `Put` is a bit weird. Easy enough to adjust both `Inc` and `Put`, and maybe `Scan`, to not be variadic.

bdarnell commented on client/db.go:
```diff
@@ -0,0 +1,461 @@
+
+// ValueString ...
+func (kv *KeyValue) ValueString() string {
+	return string(kv.value.([]byte))
```
This will always copy, which seems unfortunate. If we offer both ValueString and ValueBytes, one of them will always copy (no matter which one we pick as the "native" value), so I think we should probably pick only one or the other. The application can do the conversion if they need to, but then it will be clear which option incurs the cost of a copy.

tschottdorf replied:
what if we have ValueStringer() that simply returns `kv.value`, which we turn into a Stringer (returning the value as the string)?

Then everything that works on strings and properly abstracts should be able to work with it pretty naturally, and it's clear that the bytes are the option which is more natural to the KeyValue struct.

petermattis replied:
`kv.value` is itself an `interface{}` so you can't implement another interface on it. Ben is probably correct that only have one method (probably `ValueBytes`) and requiring the caller to cast to `string` is better.

tschottdorf replied:
but you're type asserting the interface away, so you can do it: http://play.golang.org/p/rNvRUBgS_4
If I'm using the client with a codebase that passes everything as stringer (or just uses printf a lot) I think I would prefer that to having to cast everywhere, but then again how often will that happen.

petermattis replied:
Oh, good point. That would help if the result from `ValueStringer()` is passed as an `interface{}`, but if someone wants a `string` the cast will be simpler. Anyways, I've added: folks can see what it looks like.

spencerkimball commented on client/db.go:
```diff
@@ -0,0 +1,461 @@
+	return fmt.Errorf("unable to unmarshal value: %T -> %T", kv.value, v)
+}
+
+func (kv *KeyValue) setValue(v *proto.Value) {
```
There's no point in being allowed to set the timestamp on a value before putting it to the database. This will be disallowed if it doesn't match the timestamp of the actual PutRequest.

petermattis replied:
Not sure I'm understanding this comment. This method is usually used to populate the value portion of `KeyValue` from a response.

spencerkimball replied:
Oh then i misunderstood it. Ignore.

bdarnell replied:
Why go through this conversion instead of simply storing the proto.Value as a field of KeyValue?

petermattis replied:
Not sure. Let me see if that works.

tschottdorf commented:
What would batching up and executing a bunch of different calls look like?

spencerkimball commented on client/db.go:
```diff
@@ -0,0 +1,461 @@
+		log.Fatal(err)
+	}
+	kv := NewKV(nil, sender)
+	kv.User = "root"
```
TODO here.

petermattis replied:
Yeah. I'm thinking that `addr` should be some sort of URI. Something like `(http|https|rpc|rpcs)://<user>@<hostname>:<port>`.

spencerkimball commented on client/db.go:
```diff
@@ -0,0 +1,461 @@
+// Scan ...
+func (b *Batch) Scan(maxRows int64, args ...interface{}) error {
+	if len(args)%2 != 0 {
+		return fmt.Errorf("Put requires pairs of arguments, got %d", len(args))
```
s/Put/Scan/

petermattis replied:
Oops.

spencerkimball commented on client/db_test.go:
```diff
@@ -0,0 +1,28 @@
+	fmt.Println(result.Rows[0].ValueString())
+	fmt.Println(result.Rows[1].ValueInt())
+
+	// Output:
```
Could you add a batch, a txn, and a txn that uses a batch as examples?

petermattis replied:
Sure. Ran out of time before I had to leave today.

spencerkimball commented on client/db_test.go:
```diff
@@ -0,0 +1,28 @@
+	defer s.Stop()
+
+	db := client.Open(s.ServingAddr())
+	db.Put("hello", "world")
```
This is a little confusing...using "world" as a value and a key. I'd rejigger this slightly.

spencerkimball commented:
This is cool. A couple more comments:

- We need to be careful about the namespace in client/ because we're going to be adding a structured client. Perhaps we should be moving this stuff into client/kv and then just sucking it up in the server code and naming the kv import something different if it conflicts. The key aim here should be to make things understandable to end users.
- Do you envision keeping this layering? Or should it replace the stuff we have now entirely? I want to keep the implementation burden straightforward for other clients.

bdarnell commented:
In general I like the direction; it's nice to have something a little more tailored to the core KV methods than the generic request/response stuff. The use of `interface{}` is unfortunate but I guess the only alternative is to pick a single type.

petermattis commented:
@spencerkimball Not sure about the namespace and how this interacts with the structured data API. Let's figure out the location after the the interface. I think this needs to layer (or live alongside) the existing `client` stuff just for ease of development and transition. My previous experience with making sweeping changes to the `client` package is that we have to do it in steps. 

It might be worthwhile to tailor the protocol to more efficiently implement the interface. For example, in this API everything is a `Batch` underneath the hood. Perhaps `Batch{Request,Response}` is the only endpoint that the server needs to implement and that the the format of requests and responses can minimize allocations. I'm not proposing to do this now, just pointing out the possibility. The interface is the important part right now.

petermattis commented:
Expanded `ExampleDB` to show the use of a transaction and a batch containing a `Get`, `Scan` and `Put`. I believe I've addressed all of the comments. PTAL.

tschottdorf commented on client/db_test.go:
```diff
@@ -0,0 +1,75 @@
+	if err := db.Run(b); err != nil {
+		log.Fatal(err)
+	}
+	fmt.Println(4, b.Results[0].Rows[0].ValueStringer())
```
cool. This looks pretty smooth already. This part would probably bloat in actual code:

```
if len(b.Rows) > 0 { ... }
```

Maybe something like

`val, ok := b.Results[0].ValueStringer()`?
`i, ok := b.Results[1].ValueInt()`?

to get the first result (if there is one), and for other indexes
`val_i, ok := b.ValueInt(i)`

and for slices

`val_two_and_three_as_slice, ok := b.Results.ValueBytes(2, 3)`

Just brainstorming here.

tschottdorf replied:
Lookup by key could also be interesting sometimes:
```
_ := b.Get(keyA, keyB, keyC).Run()
v, ok := b.Results.ValueBytes(keyB)
```

petermattis replied:
How would the lookup by key be implemented? Scanning over the results? Seems bad to provide an interface that can't be implemented efficiently when the user can easily keep track of which row will have the correct key.

I'm not sure about the other suggestion regarding accessors for the results. It isn't called out in the non-existent documentation, but for `Get`, `Put`, `Inc` and `Del` there is well defined number of rows that will be returned. The docs should eventually say something like: Each operation (`Get`, `Put`, `Inc`, `Scan` or `Del`) will add a new `Result` to the batch. For `Get` and `Del`, the `Result` will contain exactly as many rows as the number of arguments passed to `Get` and `Del`. For `Put` and `Inc`, the `Result` will contain exactly 1 row. For `Scan` the number of rows is variable but is capped at the supplied `maxRows` parameter.

tschottdorf commented on client/db_test.go:
```diff
@@ -0,0 +1,75 @@
+		log.Fatal(err)
+	}
+	result = db.Get("key", "aa", "inc")
+	fmt.Println(9, result.Rows[0].Exists())
```
this would be handled nicely with my above api suggestion. If it exists, most of the time you also want the value.

tschottdorf commented on client/db.go:
```diff
@@ -0,0 +1,473 @@
+}
+
+// Get ...
+func (b *Batch) Get(args ...interface{}) {
```
Batch could be more useful if it returned itself, then you could run

`err := b.Get(...).Put(...).Scan(...).Run()`

spencerkimball commented on client/db.go:
```diff
@@ -0,0 +1,529 @@
+		switch t := v.(type) {
+		case encoding.BinaryUnmarshaler:
+			return t.UnmarshalBinary(val)
+		case gogoproto.Message:
```
What are we going to do about other protobuf implementations which people are more likely to use than gogoproto? It's a shame that protobuf doesn't support UnmarshalBinary. Am I missing why they couldn't / don't?

bdarnell replied:
I think I'd rather see Value take a BinaryUnmarshaler instead of an `interface{}` (since we already have ValueBytes() for convenient raw access). With `interface{}` it's too easy to make pointer versus non-pointer mistakes, and those mistakes can't be caught until runtime.

We could provide a little shim to adapt a gogoproto.Message to a BinaryUnmarshaler (I don't know why they don't implement it themselves. It doesn't seem commonly used; I wasn't aware of it before), and this would also be more extensible to other formats.

petermattis replied:
Or perhaps `Value` should take a `BinaryUnmarshaler` but we add a `ValueProto` for retrieving protos.

bdarnell commented on client/db.go:
```diff
@@ -0,0 +1,529 @@
+}
+
+// ValueStringer ...
+func (kv *KeyValue) ValueStringer() fmt.Stringer {
```
This seems awkward to me. The name `Stringer`, while standard, is not seen very often in practice, so it feels unfamiliar. It's better to say `string(kv.ValueBytes())` than `kv.ValueStringer.String()`. This is useful when you don't know the underlying type, but when would you need to be type-agnostic but also care about the efficiency of Stringer vs string?

petermattis replied:
Yeah, I was just experimenting with Toby's suggestion. Doesn't seem worthwhile.

tschottdorf replied:
It's nice for Printf, but doesn't seem too valuable otherwise. Toss it.

spencerkimball commented on client/db_test.go:
```diff
@@ -0,0 +1,69 @@
+	"github.com/cockroachdb/cockroach/server"
+)
+
+func ExampleDB() {
```
I was suggesting creating several examples. The original one was so small and concise, which gave it real appeal. I think now it's unnecessarily murky.

petermattis replied:
Ok, I broken this up into multiple smaller examples.

tschottdorf commented on client/db_test.go:
```diff
@@ -0,0 +1,130 @@
+	if err := db.Run(b); err != nil {
+		log.Fatal(err)
+	}
+	fmt.Println(b.Results[0].Rows[0].String())
```
just in case you didn't see them, I made some comments about this part in https://github.com/cockroachdb/cockroach/pull/871#discussion-diff-29486757

tschottdorf commented:
For the key lookup you could build a translation table `key->index` but let's forget about this for now.

Especially for operations where you know that there's exactly one thing it's awkward that I have to write `.Rows[0].` all the time. Just having to do `result.ValueBytes()` seems like an improvement. And sure, I can check manually if that thing is nil if I care whether there actually was a value, but if I can just write `val, ok := ...` I'll usually have exactly what I care about in one statement.

And then, if a `Get(a,b,c)` gives you a result with three rows, you might want to do that with the second row and just get the same info.

The slice accessor, well, I don't feel too strongly about that one. Probably much less useful.

petermattis commented:
Constructing the `key->index` translation table will penalize users who don't use it unless we only build it on demand. And what to do if there are multiple identical keys? For example, `Get("a", "a")`. Yeah, users shouldn't do that, but they will.

Would `result.ValueBytes()` return the bytes for row `0`? Maybe that should be `result.ValueBytes(n)`.

Bigger question: how should timestamps be specified for requests? One option would be to add `At` variants of the ops: `GetAt`, `ScanAt`, etc. Seems excessive. Another option would be to only allow specifying timestamps at the batch level. That is, add `Batch.Timestamp` and propagate that timestamp to all of the requests.

petermattis commented:
Alright, this is ready for another look. I stripped out the error handling from the examples to keep the neat and tidy. I've added support for conditional put (`CPut`) and delete range (`DelRange`). Any thoughts on my earlier question of specifying timestamps in requests?

tschottdorf commented:
Why strip the error handling? An example with "real" production-grade error handling would be nice. In the end, people are supposed to do things correctly.

petermattis commented:
The production-grade error handling is verbose, though not horribly so. I'm open to putting it back it, but think it looks nicer without it.

spencerkimball commented on client/db.go:
```diff
@@ -0,0 +1,627 @@
+// Timestamp ...
+func (kv *KeyValue) Timestamp() time.Time {
+	sec := kv.timestamp / 1e9
+	nsec := kv.timestamp - sec*1e9
```
Why not just do kv.timestamp % 1e9

petermattis replied:
Habit. The above is faster, though obviously it doesn't matter for this code. I'll change it.

spencerkimball commented:
LGTM

tschottdorf commented:
@petermattis 
> Would result.ValueBytes() return the bytes for row 0? Maybe that should be result.ValueBytes(n).

What do you mean `(n)`? I'd use `result.ValueBytes()` as sugar for `result.ValueBytes(0)`. But I'm okay with leaving that out and just having `result.ValueBytes(i)` (though I think the sugar is still worth it). My main point is the ease of access and checking if there's a value via the `(val, ok)` signature.

tschottdorf commented:
Specifying timestamps is useful almost exclusively for the reason of causality, right (first txn finishes, and you want to run another one and pass that timestamp in because you definitely want to make sure that your second txn gets a later commit timestamp)? That seems pretty rare so only having it on batch as a sidenote seems fine.

petermattis commented:
Added a bunch of documentation comments. The new API needs more testing, but I think this PR should land without it. My plan is to migrate uses of the old client API over to this new one, including all of the tests.

spencerkimball commented:
Yeah, land it. LGTM
